exposed-core/src/main/kotlin/org/jetbrains/exposed/sql/Query.kt

```diff
@@ -77,9 +87,43 @@ open class Query(override var set: FieldSet, where: Op<Boolean>?) : AbstractQuer
     }
 
     override fun withDistinct(value: Boolean): Query = apply {
+        if (value) {
+            require(distinctOn == null) { "DISTINCT cannot be used with the DISTINCT ON modifier. Only one of them should be applied." }
+        }
         distinct = value
     }
 
+    /**
+     * Specifies that the `SELECT` query should retrieve distinct results based on the given list of columns.
+     *
+     * This method can be used to set a `DISTINCT ON` clause for the query, which is supported by some SQL dialects
+     * (e.g., PostgreSQL, H2). The resulting query will retrieve rows that are distinct based on the specified columns.
+     *
+     * @param columns The columns to apply the `DISTINCT ON` clause.
+     * @return The current `Query` instance with the `DISTINCT ON` clause applied.
+     */
+    fun withDistinctOn(vararg columns: Column<*>): Query = apply {
+        require(!distinct) { "DISTINCT ON cannot be used with the DISTINCT modifier. Only one of them should be applied." }
+        distinctOn = (distinctOn ?: emptyList()) + columns
+    }
+
+    /**
+     * Specifies that the `SELECT` query should retrieve distinct results based on the given list of columns with sort orders.
+     * This method sets a `DISTINCT ON` clause and may reorder the results as indicated.
+     *
+     * This method can be used to set a `DISTINCT ON` clause for the query, which is supported by some SQL dialects
+     * (e.g., PostgreSQL, H2), along with an `ORDER BY` clause for the specified columns.
+     *
+     * @param columns The columns and their sort orders to apply the `DISTINCT ON` clause.
+     * @return The current `Query` instance with the `DISTINCT ON` clause and reordering applied.
+     */
+    fun withDistinctOn(vararg columns: Pair<Column<*>, SortOrder>): Query = apply {
+        require(!distinct) { "DISTINCT ON cannot be used with the DISTINCT modifier. Only one of them should be applied." }
+        @Suppress("SpreadOperator")
+        withDistinctOn(*columns.map { it.first }.toTypedArray())
+        return orderBy(*columns)
```

Comment:
(minor) with this implementation it would currently be allowed to -- arguably unintentionally -- add an entry to `orderByExpressions` twice for a DISTINCT ON column (since that attribute is a list and not a distinct set), e.g. 
```
table
  .selectAll()
  .withDistinctOn(table.col to SortOrder.ASC)
  .orderBy(table.col, SortOrder.DESC)
```
postgresql allows for this and uses [the last specified sorting predicate specified for a column](https://www.postgresql.org/docs/current/queries-order.html#:~:text=When%20more%20than%20one%20expression%20is%20specified%2C%20the%20later%20values%20are%20used%20to%20sort%20rows%20that%20are%20equal%20according%20to%20the%20earlier%20values.) but i'm not sure if you're intentionally supporting this use case; i don't personally know if the behaviour of having a column mentioned twice in `ORDER BY` is consistent across all sql dialects.

Reply:
Thank you for the feedback,

when I was adding the option of `withDistinctOn` with ordering I also thought that it could cause misunderstanding,

In general, the main purpose of that variant is make a use case when there is only distinctOn without extra ordering simpler, and prevent necessity to call `orderBy` right after `withDistinctOn()` with the same columns order, what could be broken on any refactoring.

So my expectation at the current moment that for any complex logic of adding ordering, the version `withDistinctOn(vararg columns: Column<*>)` (without ordering) should be used.